[FSDP] Propagate requires_grad attribute to unsharded params

edpizzi opened the pull request:
Summary:
This preserves `requires_grad` in the case where all parameters within a `FlatParameter` have the same `requires_grad` value.

Currently, unsharded parameters have `requires_grad=True` in some cases where the `FlatParameter` and all original parameters have `requires_grad=False`.

This could be extended to support `FlatParameters` with a mix of `requires_grad` states by extending `ParamInfo` to capture `requires_grad` for each parameter.

Test Plan: test added

Differential Revision: D49517155

awgu reviewed (approved):
This makes sense to me!

edpizzi committed:
[FSDP] Propagate requires_grad attribute to unsharded params (#109892)

Summary:

This preserves `requires_grad` in the case where all parameters within a `FlatParameter` have the same `requires_grad` value.

Currently, unsharded parameters have `requires_grad=True` in some cases where the `FlatParameter` and all original parameters have `requires_grad=False`.

This could be extended to support `FlatParameters` with a mix of `requires_grad` states by extending `ParamInfo` to capture `requires_grad` for each parameter.

Test Plan: test added

Reviewed By: awgu

Differential Revision: D49517155

edpizzi commented:
@pytorchbot merge

edpizzi commented:
@pytorchbot merge
